File: app/src/main/java/com/kjjejones42/audiobookplayer/display/DisplayListViewModel.java
package com.kjjejones42.audiobookplayer.display;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.kjjejones42.audiobookplayer.AudioBook;
import com.kjjejones42.audiobookplayer.Utils;
import com.kjjejones42.audiobookplayer.database.AudiobookDatabase;

import java.util.ArrayList;
import java.util.List;

public class DisplayListViewModel extends ViewModel {

    private final MutableLiveData<List<AudioBook>> books = new MutableLiveData<>(new ArrayList<>());
    private final MutableLiveData<List<ListItem>> listItems = new MutableLiveData<>(new ArrayList<>());

    public DisplayListViewModel() {
    }

    void loadFromDisk(Context context) {
        try {
            books.setValue(AudiobookDatabase.getInstance(context).audiobookDao().getAll());
            listItems.setValue(listItems.getValue());
        } catch (Exception e) {
            Utils.logError(e, context);
            e.printStackTrace();
        }
    }

    @NonNull
    LiveData<List<AudioBook>> getSavedBooks(Context context) {
        if (books.getValue() == null) {
            loadFromDisk(context);
        }
        return books;
    }

    @NonNull
    LiveData<List<ListItem>> getListItems(Context context) {
        if (listItems.getValue() == null) {
            loadFromDisk(context);
        }
        return listItems;
    }

    void setListItems(List<ListItem> newList) {
        listItems.setValue(newList);
    }
}

File: app/src/main/java/com/kjjejones42/audiobookplayer/display/DisplayListViewModel.kt
package com.kjjejones42.audiobookplayer.display

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.work.WorkInfo
import androidx.work.WorkManager
import com.kjjejones42.audiobookplayer.AudioBook
import com.kjjejones42.audiobookplayer.database.AudiobookRepository
import com.kjjejones42.audiobookplayer.display.ListItem.AudioBookContainer
import com.kjjejones42.audiobookplayer.display.ListItem.Heading
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import java.util.UUID
import javax.inject.Inject

@HiltViewModel
class DisplayListViewModel @Inject constructor(
    audiobookRepository: AudiobookRepository
) : ViewModel() {

    val listItems = audiobookRepository
        .allAndObserve()
        .map { getItemsFromBooks(it) }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = emptyList()
        )

    val workUuid = MutableStateFlow<UUID?>(null)

    val workNeedsStarting = MutableStateFlow(false)

    fun getWorkStateFlow(context: Context): Flow<WorkInfo?> {
        val uuid = workUuid.value ?: return flowOf(null)
        return WorkManager.getInstance(context).getWorkInfoByIdFlow(uuid)
    }

    companion object {
        fun getItemsFromBooks(books: List<AudioBook>): List<ListItem> {
            val list = books
                .sortedBy { it.displayName }
                .map { AudioBookContainer(it) }
                .toMutableList<ListItem>()

            list.toList()
                .map { it.category }
                .distinct()
                .map { Heading(it) }
                .forEach { list.add(it) }

            return list.sortedWith(compareBy({ it.category }, { it.type.value }, { -it.timeStamp }))
        }
    }
}